### PR TITLE
fix: issues with invoices

### PR DESCRIPTION
**Pull request by Cameri**
## Types of changes

- [ ] Non-functional change (docs, style, minor refactor)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my code changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: resources/invoices.html [hunk 1]
```diff
@@ -164,14 +168,23 @@ <h2 class="text-danger">Invoice expired!</h2>
             case 'EVENT': {
               // TODO: validate event
               const event = message[2]
+              console.log('event', event)
```
Comment (antonleviathan) [inline resources/invoices.html:171]: Do we need this log?

File: resources/invoices.html [hunk 2]
```diff
@@ -164,14 +168,23 @@ <h2 class="text-danger">Invoice expired!</h2>
             case 'EVENT': {
               // TODO: validate event
```
Comment (antonleviathan) [inline resources/invoices.html:169]: Is this an old comment?

File: src/app/maintenance-worker.ts [hunk 3]
```diff
@@ -47,21 +47,26 @@ export class MaintenanceWorker implements IRunnable {
...
           invoice.status !== updatedInvoice.status
           && updatedInvoice.status == InvoiceStatus.COMPLETED
-          && invoice.confirmedAt
+          && updatedInvoice.confirmedAt
         ) {
           debug('confirming invoice %s & notifying %s', invoice.id, invoice.pubkey)
+          const update = {
+              ...invoice,
+              ...{ amountPaid: invoice.amountRequested },
+              ...updatedInvoice,
```
Comment (antonleviathan) [inline src/app/maintenance-worker.ts:65]: Should we use a [`deepMerge`](https://ramdajs.com/docs/#mergeDeepLeft) here?

Comment (Cameri) [reply src/app/maintenance-worker.ts]: done

File: resources/invoices.html (hunk 2 above)
Comment (Cameri) [reply resources/invoices.html]: indeed

File: src/handlers/event-message-handler.ts [hunk 4]
```diff
@@ -249,6 +261,10 @@ export class EventMessageHandler implements IMessageHandler {
       return
     }
 
+    if (this.getRelayPublicKey() === event.pubkey) {
+      return undefined
```
Comment (antonleviathan) [inline src/handlers/event-message-handler.ts:265]: Inconsistent returns, is this intentional?

File: src/services/payments-service.ts [hunk 5]
```diff
@@ -309,31 +245,33 @@ ${invoice.bolt11}`,
...
     }
 
-    const persistEvent = this.eventRepository.create.bind(this.eventRepository)
+    const persistEvent = async (event: Event) => {
+      await this.eventRepository.create(event)
+
+      return event
+    }
 
     await pipe(
```
Comment (antonleviathan) [inline src/services/payments-service.ts:270]: Should we be catching here for errors?

File: src/handlers/event-message-handler.ts (hunk 4 above)
Comment (Cameri) [reply src/handlers/event-message-handler.ts:265]: this is fine, we return `undefined` if user is admitted/whitelisted or a string with the reason if they are not admitted

this new check fixes the issue of the relay's public key not being allowed on the relay when payment is enabled which is a big oversight

File: src/utils/event.ts [hunk 6]
```diff
@@ -181,15 +181,33 @@ export const identifyEvent = async (event: UnidentifiedEvent): Promise<UnsignedE
   return { ...event, id }
 }
 
-export function getRelayPrivateKey(relayUrl: string): string {
+let privateKeyCache: string | undefined
+export function getRelayPrivateKey(secret: string): string {
```
Comment (antonleviathan) [inline src/utils/event.ts:185]: Is `secret` an optional parameter? If so we could make the type indicate this

File: src/utils/event.ts [hunk 7]
```diff
@@ -181,15 +181,33 @@ export const identifyEvent = async (event: UnidentifiedEvent): Promise<UnsignedE
...
 }
 
-export const getPublicKey = (privkey: string | Buffer) => Buffer.from(secp256k1.getPublicKey(privkey, true)).subarray(1).toString('hex')
+const publicKeyCache: Record<string, string> = {}
+export const getPublicKey = (privkey: string) => {
+  if (privkey in publicKeyCache) {
+    return publicKeyCache[privkey]
+  }
+
+  publicKeyCache[privkey] = Buffer.from(secp256k1.getPublicKey(privkey, true)).subarray(1).toString('hex')
```
Comment (antonleviathan) [inline src/utils/event.ts:207]: You can do something like this instead:
`secp256k1.utils.bytesToHex(secp256k1.getPublicKey(privkey, true)).subarray(1))`

File: src/utils/event.ts (hunk 6 above)
Comment (Cameri) [reply src/utils/event.ts:185]: it should always be the relay's URL, but it is ignored when `RELAY_PRIVATE_KEY` env var is set

File: src/handlers/event-message-handler.ts (hunk 4 above)
Comment (Cameri) [reply src/handlers/event-message-handler.ts:265]: in the future we could change this pattern to using constants or symbols

File: src/utils/event.ts (hunk 6 above)
Comment (antonleviathan) [reply src/utils/event.ts:185]: Ok, why is the param called `secret`?

File: src/services/payments-service.ts (hunk 5 above)
Comment (Cameri) [reply src/services/payments-service.ts:270]: sure! will add a catch

File: src/handlers/event-message-handler.ts (hunk 4 above)
Comment (Cameri) [reply src/handlers/event-message-handler.ts:265]: Ah, I'll remove the extra undefined

**Pull request merged**